### PR TITLE
[FIX] tools: fix filename resolution in XSD imports

### PR DESCRIPTION
**Pull request by antoine162**
With PRs odoo/#89145 and [enterprise/#26393](https://github.com/odoo/enterprise/pull/26393) came new methods for retrieving XSD files and using them for XML validation.

The new retrieval method expects modules to provide a 'prefix' that is prepended to the XSD's filename. For example, l10n_cl_edi will name its XSD files 'l10n_cl_edi.filename.xsd'.

However, this messes things up when one XSD file needs to import another. For example, l10n_cl_edi.DTE_v10.xsd has the statement '<xs:include schemaLocation="SiiTypes_v10.xsd"/>'

Currently, the filename resolver has no way of knowing that this should resolve to 'l10n_cl_edi.SiiTypes_v10.xsd', not 'SiiTypes_v10.xsd'.

In addition, the new retrieval method saves the ZIP archives received over the network under 'filename.xsd'. Thus 'SiiTypes_v10.xsd' might actually be a ZIP-encoded file.

So, we need to do something to fix the imports.
Here are two possible solutions:
1. We scrap this 'prefix' stuff so that 'SiiTypes_v10.xsd' is the correct file. Either save the ZIP files under a different name, or just don't save them.  
2. Or, we provide a mechanism for indicating a prefix to the filename resolver.

Personally, I don't see the point in saving the ZIP files, and this 'prefix' stuff seems pointless. So I prefer solution 1.

But, because I assume there must be a reason for all of that 'prefix' stuff, here is an implementation of solution 2.

I'd be keen to know the reason, btw.

EDIT: 
In addition to the first issue described above, we have the second issue that some XSD files returned by the Chilean SII are encoded using ISO-8859-1 encoding (e.g. SiiTypes_v10.xsd). If we leave them in this encoding, then LXML isn't able to parse them when performing imports. 

Solution: convert the files to UTF-8 before storing them.

Linked PR: https://github.com/odoo/enterprise/pull/31288
Task: 2976749

### CODE REVIEW

**Comment by jco-odoo** [comment]
@Shawcker Did you think about upgrade scripts as well?   (or does that go with without is possible?) For the encoding in iso-8859 it might be important as the XML needs to be encoded that way.  

The zip file is only there to be unzipped...

**Comment by jco-odoo** [comment]
Should not there be just one prefix?  And if multiple, should it be a list or a tuple?

**Comment by Shawcker** [comment]
> Did you think about upgrade scripts as well? (or does that go with without is possible?)

We could clean up some (old) attachments. If an xsd was already in db before, we will most likely re-download it (especially with the prefix), so outdated ones won't be used anymore (and can be removed).

> The zip file is only there to be unzipped...

Indeed... Now that I think about it, the zip file was saved as attachment to avoid having to save all the other files in it. But we decided to save them anyway, so saving the zipfile itself should not be done.

> Should not there be just one prefix? And if multiple, should it be a list or a tuple?

I'm not sure I understand what you mean

**Comment by jco-odoo** [comment]
@Shawcker I am talking about the parameter prefixes added in the PR.  I think it would be a good idea that if one xsd references another it can also check if the filename with prefix exists, but I don't see the need for multiple prefixes.  @antoine162

**Comment by Shawcker** [comment]
@jco-odoo OK, I agree we should check for filenames with a prefix. And only one prefix is enough, no need for tuple/list.

**Comment by fvz-odoo** [comment]
Hello @antoine162 and @jco-odoo 

Just tested this PR but I'm still getting the traceback... Tested invoice, boleta and delivery guide.

Let me know!

Cheers!


RPC_ERROR
Odoo Server Error
Traceback (most recent call last):
  File "/data/build/odoo/odoo/http.py", line 1519, in _serve_db
    return service_model.retrying(self._serve_ir_http, self.env)
  File "/data/build/odoo/odoo/service/model.py", line 134, in retrying
    result = func()
  File "/data/build/odoo/odoo/http.py", line 1548, in _serve_ir_http
    response = self.dispatcher.dispatch(rule.endpoint, args)
  File "/data/build/odoo/odoo/http.py", line 1745, in dispatch
    result = self.request.registry['ir.http']._dispatch(endpoint)
  File "/data/build/odoo/addons/website/models/ir_http.py", line 221, in _dispatch
    response = super()._dispatch(endpoint)
  File "/data/build/odoo/odoo/addons/base/models/ir_http.py", line 138, in _dispatch
    result = endpoint(**request.params)
  File "/data/build/odoo/odoo/http.py", line 673, in route_wrapper
    result = endpoint(self, *args, **params_ok)
  File "/data/build/odoo/addons/web/controllers/dataset.py", line 46, in call_button
    action = self._call_kw(model, method, args, kwargs)
  File "/data/build/odoo/addons/web/controllers/dataset.py", line 33, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/data/build/odoo/odoo/api.py", line 461, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "/data/build/odoo/odoo/api.py", line 448, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/data/build/enterprise/l10n_be_reports_post_wizard/models/account_move.py", line 36, in action_post
    return super().action_post()
  File "/data/build/odoo/addons/sale/models/account_move.py", line 63, in action_post
    res = super(AccountMove, self).action_post()
  File "/data/build/odoo/addons/account/models/account_move.py", line 3501, in action_post
    other_moves._post(soft=False)
  File "/data/build/enterprise/l10n_cl_edi/models/account_move.py", line 142, in _post
    move._l10n_cl_create_dte()
  File "/data/build/enterprise/l10n_cl_edi/models/account_move.py", line 486, in _l10n_cl_create_dte
    signed_dte = self._sign_full_xml(
  File "/data/build/enterprise/l10n_cl_edi/models/l10n_cl_edi_util.py", line 257, in _sign_full_xml
    self._xml_validator(signature, 'sig')
  File "/data/build/enterprise/l10n_cl_edi/models/l10n_cl_edi_util.py", line 224, in _xml_validator
    return tools.validate_xml_from_attachment(self.env, xml_to_validate, xsd_fname, self.env['ir.attachment']._l10n_cl_edi_load_xsd_files)
  File "/data/build/odoo/odoo/tools/xml_utils.py", line 278, in validate_xml_from_attachment
    _check_with_xsd(xml_content, xsd_name, env)
  File "/data/build/odoo/odoo/tools/xml_utils.py", line 56, in _check_with_xsd
    xsd_schema = etree.XMLSchema(etree.parse(stream, parser=parser))
  File "src/lxml/etree.pyx", line 3521, in lxml.etree.parse
  File "src/lxml/parser.pxi", line 1876, in lxml.etree._parseDocument
  File "src/lxml/parser.pxi", line 1896, in lxml.etree._parseMemoryDocument
  File "src/lxml/parser.pxi", line 1784, in lxml.etree._parseDoc
  File "src/lxml/parser.pxi", line 1141, in lxml.etree._BaseParser._parseDoc
  File "src/lxml/parser.pxi", line 615, in lxml.etree._ParserContext._handleParseResultDoc
  File "src/lxml/parser.pxi", line 725, in lxml.etree._handleParseResult
  File "src/lxml/parser.pxi", line 654, in lxml.etree._raiseParseError
  File "<string>", line 1
lxml.etree.XMLSyntaxError: Start tag expected, '<' not found, line 1, column 1

**Comment by fvz-odoo** [comment]
@jco-odoo ping

**Comment by jco-odoo** [comment]
@Shawcker Could you have a last look as well?

**Comment by jco-odoo** [comment]
@antoine162 will still try to remove the validate_xsd_content param and reuse the modify one.

File: odoo/tools/xml_utils.py [hunk 1]
```diff
@@ -20,18 +20,23 @@ class odoo_resolver(etree.Resolver):
...
+        self.prefix = prefix
 
     def resolve(self, url, id, context):
         """Search url in ``ir.attachment`` and return the resolved content."""
+        if self.prefix:
+            attachment = self.env['ir.attachment'].search([('name', '=', f'{self.prefix}.{url}')])
+            if attachment:
+                return self.resolve_string(attachment.raw, context)
         attachment = self.env['ir.attachment'].search([('name', '=', url)])
         if attachment:
```
Comment (Shawcker) [inline odoo/tools/xml_utils.py:35]:
```suggestion
        attachment_name = f'{self.prefix}.{url}' if self.prefix else url
        attachment = self.env['ir.attachment'].search([('name', '=', attachment_name)])
        if attachment:
            return self.resolve_string(attachment.raw, context)
```

File: odoo/tools/xml_utils.py [hunk 2]
```diff
@@ -203,6 +212,8 @@ def load_xsd_files_from_url(env, url, file_name, force_reload=False,
     content = response.content
     if modify_xsd_content and not is_zip:
         content = modify_xsd_content(content)
+    if validate_xsd_content and not is_zip:
+        content = etree.tostring(objectify.fromstring(content), encoding='utf-8', pretty_print=True)
```
[the remark this answers is not in the document]
Comment (antoine162) [reply odoo/tools/xml_utils.py]: yes indeed, I've done that :+1:

File: odoo/tools/xml_utils.py (hunk 1 above)
Comment (antoine162) [reply odoo/tools/xml_utils.py:34]: yes, I agree this is better :+1:

File: odoo/tools/xml_utils.py [hunk 3]
```diff
@@ -148,7 +152,7 @@ def leaf_iter(parent_node, node, level):
 
 
 def load_xsd_files_from_url(env, url, file_name, force_reload=False,
-                            request_max_timeout=10, xsd_name_prefix='', xsd_names_filter=None, modify_xsd_content=None):
+                            request_max_timeout=10, xsd_name_prefix='', xsd_names_filter=None, modify_xsd_content=None, validate_xsd_content=False):
```
Comment (jco-odoo) [inline odoo/tools/xml_utils.py:155]: @antoine162

Comment (antoine162) [reply odoo/tools/xml_utils.py]: ahhh yes, I've removed that

**Review by odony — COMMENTED** [review]
That seems reasonable.

By the way the prefix is going to be helpful to avoid collisions with other files, but this is still not a robust design. The lookup of attachments by only their `name` means things will blow up as soon as someone creates an attachment with the same name *anywhere* in the database. It could be an attachment on a ticket, on a job application, a file uploaded in Documents, etc.

This could be made more robust with other attributes of attachments, e.g. `url`, `type`, `create_uid`, etc, to make sure the XSD attachments can't be mixed up with other attachments. And of course, handle the case where multiple attachments do eventually match, instead of crashing.

**Pull request closed without merging**